### ISSUE TITLE
Refactor adapter protocol

### ISSUE THREAD

**Issue by vanstee** [operable/cog#288]
Now that we have two usable adapter implementations (slack and hipchat) it'd be nice to take a look at the adapter protocol to remove unused callbacks and add callbacks for common functions. More details to follow after I review the current state of things.

**Comment by vanstee** [comment]
We currently have `describe_tree/1`, `message/2`, `lookup_room/1` and `lookup_user/1` but it seems like we're really only using `describe_tree/1` and `lookup_room/1` (externally to the adapter). Here's what I plan on fixing with this PR:

* Remove `lookup_user/1` in favor of allowing adapters to use their own api functions internally.
* Make the executor and other modules call `send_message/2` (renamed from `message/2`) and `lookup_room/1` defined on the adapter instead of any submodules.
* Add a generic hook for adapters to put messages on the bus, since right now that's all handled in random places in the rim connector areas.
* Add a generic way to cache api calls.

To help make sure this is sane, it might be nice to implement an IRC adapter alongside this work to make sure this API works with a significantly different service.

**Comment by christophermaier** [comment]
Why remove `lookup_user/1`? That seems like core common functionality to define in a behavior. The implementations would be free to do this however they want, of course.

Also see `mention_name/1` and `service_name/1` in the current Slack and Hipchat API modules.

**Comment by vanstee** [comment]
@christophermaier It seems useful, but I couldn't actually find where we needed that function external to an adapter. But you're right about the others:

* Add `service_name/0` to defined a name for the adapter
* Add `lookup_user/1`
* Maybe add a way to pull history of messages which is currently only used in integration tests.

**Comment by christophermaier** [comment]
@vanstee Ah, I think I got turned around a bit; I was thinking it was needed for resolving redirects, but that's using `lookup_room/1`, but also `lookup_direct_room/1` (which wasn't on your list, but could perhaps be folded into `lookup_room/1`)

**Comment by kevsmith** [comment]
A few questions & comments:

1. What does a generic message publishing hook look like? I'm wary of adding generic code targeting future use cases unless we have a clear idea of what those use cases are.
1. Can you speak a bit more about the generic cache? I'm curious what the implementation might look like.
1. While we're in the code we should replace use of Erlang message sends (GenServer calls and the like) with MQTT.
1. We should also think about a way to describe adapter capabilities in this API. We can avoid coding to the lowest common denominator and gracefully degrade functionality. For example, let's say IRC doesn't have the concept of a direct message. Cog could query the adapter capabilities and inform the user when they attempt a redirect to a individual user.

Very rough sketch of what the code _could_ look like:

```elixir
defmodule Cog.AdapterCapability do
  defstruct name: nil, enabled: false

  @type capability_name() :: direct_message | presence

  @type t() :: %AdapterCapability {
    name :: capability_name(),
    enabled :: boolean()
  }
end

defmodule Cog.Adapter do
  alias Cog.AdapterCapability
  @type lookup_result() :: {:ok, String.t} | nil | {:error, any}
  @type lookup_opts() :: [id: String.t] | [name: String.t]

  @callback describe_tree() :: [Supervisor.Spec.spec] | []
  @callback send_message(room :: String.t, message :: String.t) :: :ok | :error
  @callback lookup_room(lookup_opts()) :: lookup_result()
  @callback lookup_user(lookup_opts()) :: lookup_result()
  @callback get_capabilities() :: [AdapterCapability.t()]
end
```

**Comment by vanstee** [comment]
@christophermaier I was planning on rolling that into `lookup_room` but now with @kevsmith proposal for capabilities, it might be more clear if we keep it separate as `lookup_direct_room`.

**Comment by vanstee** [comment]
@kevsmith I'll have some POC examples of the publishing callback and caching stuff in a bit. For the switch to using MQTT for everything, I was hoping the adapters wouldn't need to know about the message queue and would just send erlang messages back and forth, then only the top level adapter would use the publishing callback to put a message on the queue. :+1: to the capabilities idea, I'll get a POC working for that as well.

**Comment by kevsmith** [comment]
Converting to MQTT is probably worth its own issue anyways so I'm happy to drop it for now.

**Comment by vanstee** [comment]
Decided to wait on the generic caching bits since there were too many special cases in the slack adapter. Although I still think it'd be nice to have a `CachedGenServer` module at some point that works similar to the basic `GenServer` but just has some of the caching helper functions included.

**Comment by vanstee** [comment]
I've done most of what was discussed here in https://github.com/operable/cog/pull/299. But I'll be revisiting the adapter behavior and capability stuff as part of https://github.com/operable/cog/issues/307
